### ISSUE TITLE
Another northern temperate noon far from azimuth 180

### ISSUE THREAD

**Issue by isomeme** [shred/commons-suncalc#22]
Sorry to be the bearer of bad news again, but I just accidentally ran across an alarmingly large azimuth error for a northern temperate zone noon calculation -- it's off by more than 7 degrees from 180! As usual, here's the [test](https://github.com/isomeme/Helios/blob/master/helios/src/test/java/org/onereed/helios/sun/SunCalcTest.java) that demonstrates the problem.

My guess is that more problems will occur around the summer solstice at noon and the winter solstice at midnight in the temperate zones, as that's when the sun's azimuth angle rate gets highest. Since your algorithm works modulo one hour, it might be worthwhile to pick an hour on a date near one of those and then iterate through by small time intervals, starting calculations from each of those times and seeing what the noon or midnight errors are.

I'm going to re-implement the corrector code I had on my side a few weeks ago to stop these noon/nadir azimuth errors from impacting my app. I'd be happy to collaborate on a more thorough testing scheme to catch them earlier; I'd love to throw my corrector away at some point.

**Comment by shred** [comment]
Thank you for the report, because actually you found a stupid mistake I made.

The peaks for noon and nadir are estimated within two overlapping two-hour time slots each, giving a total window of 3 hours. The refinement is then supposed to find the true peak around the estimated time plus/minus two hours, to fully enclose the 3 hours.

However, in my code I added/subtracted `2/24.0` to the estimated time, because somehow I had assumed that I'm using days instead of hours. The window for the refinement only had a 24th of the correct size. If the peak was outside that window (which is likely), it was just rounded to the closest window limit.

In my fix, I now add/subtract the correct value of 2 hours, to get the expected 4 hours window. I also use a recursion depth of 14 for the nested intervals now, giving 2^14 = 16384 iterations within that window. Altogether, the peak value is now found with a resolution of less than a second (because 4 hours window size * 60 * 60 = 14400 seconds).

I hope this fix will now finally retire your corrector code. :smile: 

Can you build the main branch and try it with your app?

(PS: I hope the explanation was halfway understandable. My math class was in German, so I may be wrong with the English terms.)

**Comment by isomeme** [comment]
Ouch. I've been bit hard by unit errors over the years. I'm at least glad it was relatively easy to find and fix. And your explanation was considerably more than halfway understandable. ;)

I will definitely build against your main branch and let you know what happens. It's possible that won't be until this weekend, but it will be sooner if I can manage it. I'm eager to try it out, and also I don't know that git/gradle incantations involved in building against a non-release branch, so I get to learn something, too.

**Comment by isomeme** [comment]
Your master-RELEASE branch is [in a bad state](https://jitpack.io/#shred/commons-suncalc/), so I can't build with it tonight. I confirmed that my gradle config could pick up other arbitrary tags, so once I can build against your master branch we should be good to go.

**Comment by isomeme** [comment]
The logged error (still present now) is

```
ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project commons-suncalc: Fatal error compiling: invalid flag: --module-path -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by shred** [comment]
Yes, it's because jitpack is trying to compile the project with Java 8, but Java 9 is the required minimum for building.

**Comment by isomeme** [comment]
Ah, I see. The jitpack gradle dependency loader doesn't "see" master-SNAPSHOT when I try to depend on it. I know the overall setup is working because I can depend on v2-SNAPSHOT and the gradle sync works (and of course wrecks ensuing complication of my app due to its use of v3 features). Any idea what needs to be done where to get this working for master-SNAPSHOT?

**Comment by shred** [comment]
My experiments with Android are a little while ago. :wink:

You could install maven, then clone the project, change into the project directory and invoke `mvn clean install`.

After that, add `mavenLocal()` to your `build.gradle` file like this:

```
repositories {
  mavenLocal()
}
```

You should then be able to depend on `org.shredzone.commons:commons-suncalc:3.3-SNAPSHOT`.

**Comment by isomeme** [comment]
Well, that was quite a trip! Short version: Your 3.3-SNAPSHOT code produces wonderfully accurate and precise noon/nadir values for all the cases I tried. Nice!! I'll probably keep my corrector in the pipeline for now, as it adds little overhead when it's given an already-good value. But I expect to retire it within a few months.

The long version starts with "SunCalc needs JDK 9, Android Studio bundles 8, I'll just grab the latest JDK (14) and swap that in everywhere, oh dear, Gradle doesn't play well with JDK > 8..." and continues down that rathole for about 4 hours. :/

**Comment by shred** [comment]
I'm really sorry... If I knew that, I would just have sent you the SNAPSHOT jar file. Yes, building Java projects have become quite messy since JDK 9. I'm still looking for an universal way to generate javadocs on Java 9 to Java 14. :confused:

I will look into the twilight issue this weekend, and see if there is a fix. After that, I will release v3.3.

Feel free to reopen this bug if there are still issues with noon/nadir precision. I'd really like to get it resolved once and for all in v3.3. :smile:

**Comment by isomeme** [comment]
Oh, don't apologize. I had been doing a lot of things in Maven, Git and Gradle without really understanding the mechanics, and this forced me to learn a lot more about how they work. I'm sure that knowledge will be useful to me from now on. 

So far, since picking up 3.3-SNAPSHOT, I haven't seen a single noon/nadir error. In fact, the values are always at least an order of magnitude closer to 0/180 than I actually need. Very nice work!

**Comment by shred** [comment]
I've just released v3.3 with the fix. It also fixes another very old bug. If you are using `height()`, you will get different results now.

Thanks, again! :smile:
